### DISCUSSION TITLE
Question on fastapi on_event (shutdown) hooks for gunicorn-based set-up

### DISCUSSION THREAD

**Discussion by thegeekcentaur** [fastapi/fastapi#8531] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.on_event('shutdown')
async def stop_sqs_clients():
    log.debug('*** Shutting down worker thread ***')
    asyncio.ensure_future(sqs_client.close())
```

### Description

Greetings.

1. We are running fastapi with gunicorn workers inside docker container.
2. During startup we need to create an AWS sqs client to listen to a queue
3. During shutdown we want to be able to close this sqs client before doing a graceful exit
4. We have used gunicorn workers with a gunicorn_conf.py file
5. When container starts up we able to see gunicorn worker threads spawning off and each thread hosting a fastapi app, and able to hit the 'startup' event. Ex. -

@app.on_event('startup')
async def handle_start_up_sequence():
    log.debug('Handling start up sequence')
    asyncio.ensure_future(sqs_client.open())

We even printed the PID to ensure that gunicorn indeed spawned off multiple threads with fastapi instance in each.

However, when gunicorn master/parent thread sends a SIGABORT to worler threads (due to reaching timeout), we are unable to hit the on_event('shutdown') hook, thereby leaving dangling reference of an active SQS client in the worker thread, causing it to result a core dump.

We are looking for possible ways to utilize the shutdown event for gunicorn worker threads. Any leads/references/pointers/documentation would be appreciated.

Thanks in advance.

### Operating System

Linux

### FastAPI Version

0.58.1

### Python Version

Python 3.9.13

### Additional Context

gunicorn==20.0.4

**Comment by Kludex** [comment]
https://github.com/encode/uvicorn/issues/1116

**Comment by thegeekcentaur** [comment]
Thanks @Kludex for sharing this. So we still looking for a better and more elegant solution on this, right?

**Comment by Kludex** [comment]
I wouldn't say that... 🤔 I think it's a `gunicorn` issue, but there hasn't been an explanation on `gunicorn`'s side about it.

In any case, if you feel blocked, you can inherit `UvicornWorker`, and use https://github.com/encode/uvicorn/pull/1424/.

In any case, not a FastAPI issue.

**Comment by thegeekcentaur** [comment]
@Kludex 

I hear you. While going to UvicornWorker is a possibility but it is bit late in the game for our current scope of work.

Let me explore.

**Comment by Kludex** [comment]
What do you mean late? If you use gunicorn, you are using that worker anyway...

**Comment by thegeekcentaur** [comment]
I am talking about over-all testing and regression risks, purely from internal process perspectives.

**Comment by sandys** [comment]
@Kludex if i want to inherit and use it in my gunicorn+uvicornworker, how do i do this ? it seems ur patch branch was deleted ? any code exmples that u can share ?

does this work ? https://github.com/benoitc/gunicorn/issues/2339#issuecomment-1066010626

much thanks!

**Comment by Kludex** [comment]
```python
import threading

from uvicorn.workers import UvicornWorker

class CustomWorker(UvicornWorker):
    def _install_sigquit_handler(self, server: Server) -> None:
        """Workaround to install a SIGQUIT handler on workers.
        Ref.:
        - https://github.com/encode/uvicorn/issues/1116
        - https://github.com/benoitc/gunicorn/issues/2604
        """
        if threading.current_thread() is not threading.main_thread():
            # Signals can only be listened to from the main thread.
            return

        loop = asyncio.get_running_loop()
        loop.add_signal_handler(signal.SIGQUIT, self.handle_exit, signal.SIGQUIT, None)

    async def _serve(self) -> None:
        self.config.app = self.wsgi
        server = Server(config=self.config)
        self._install_sigquit_handler(server)
        await server.serve(sockets=self.sockets)
        if not server.started:
            sys.exit(Arbiter.WORKER_BOOT_ERROR)
```
Something like... I haven't tested.

**Comment by Kludex** [comment, marked as answer]
This issue was solved on uvicorn.
